packages/llvm_build.rb

```diff
@@ -0,0 +1,207 @@
+    # lldb fails on i686 due to requirement for a kernel > 4.1.
+    # See https://github.com/llvm/llvm-project/issues/57594
+    LLVM_PROJECTS_TO_BUILD = 'clang;clang-tools-extra;compiler-rt;libclc;lld;lldb;polly;pstl'.freeze
+  when 'x86_64'
+    # LLVM_TARGETS_TO_BUILD = 'X86;AMDGPU'
+    # LLVM_TARGETS_TO_BUILD = 'all'.freeze
+    @ARCH_C_FLAGS = '-fPIC'
+    @ARCH_CXX_FLAGS = '-fPIC'
+    @ARCH_LDFLAGS = ''
+    @ARCH_LTO_LDFLAGS = "#{@ARCH_LDFLAGS} -flto=thin"
+    LLVM_PROJECTS_TO_BUILD = 'clang;clang-tools-extra;compiler-rt;libclc;lld;lldb;polly;pstl'.freeze
+  end
+  @ARCH_C_LTO_FLAGS = "#{@ARCH_C_FLAGS} -flto=thin"
+  @ARCH_CXX_LTO_FLAGS = "#{@ARCH_CXX_FLAGS} -flto=thin"
+  # flang isn't supported on 32-bit architectures.
+  # openmp is its own package.
+  # LLVM_PROJECTS_TO_BUILD = 'clang;clang-tools-extra;libcxx;libcxxabi;libunwind;lldb;compiler-rt;lld;polly;openmp'.freeze
+
+  # Using Targets 'all' for non-i686 because otherwise mesa complains.
+  # This may be patched upstream as per
+  # https://reviews.llvm.org/rG1de56d6d13c083c996dfd44a32041dacae037d66
+  LLVM_TARGETS_TO_BUILD = 'all'.freeze
+
+  def self.patch
+    return unless ARCH == 'i686'
+
+    # Patch for LLVM 15 because of https://github.com/llvm/llvm-project/issues/58851
+    @llvm_i686_patch = <<~LLVM_PATCH_EOF
+      --- a/clang/lib/Driver/ToolChains/Linux.cpp	2022-11-30 15:50:36.777754608 -0500
+      +++ b/clang/lib/Driver/ToolChains/Linux.cpp	2022-11-30 15:51:57.004417484 -0500
+      @@ -314,6 +314,7 @@ Linux::Linux(const Driver &D, const llvm
+             D.getVFS().exists(D.Dir + "/../lib/libc++.so"))
+           addPathIfExists(D, D.Dir + "/../lib", Paths);
+
+      +  addPathIfExists(D, concat(SysRoot, "#{CREW_LIB_PREFIX}"), Paths);
+         addPathIfExists(D, concat(SysRoot, "/lib"), Paths);
+         addPathIfExists(D, concat(SysRoot, "/usr/lib"), Paths);
+       }
+    LLVM_PATCH_EOF
+    File.write('llvm_i686.patch', @llvm_i686_patch)
+    system 'patch -Np1 -i llvm_i686.patch'
+  end
+
+  def self.build
+    ############################################################
+    puts "Building LLVM Targets: #{LLVM_TARGETS_TO_BUILD}".lightgreen
+    puts "Building LLVM Projects: #{LLVM_PROJECTS_TO_BUILD}".lightgreen
+    ############################################################
+
+    unless Dir.exist?('builddir')
+      FileUtils.mkdir_p 'builddir'
+      system "echo '#!/bin/bash
+machine=$(gcc -dumpmachine)
+version=$(gcc -dumpversion)
+gnuc_lib=#{CREW_LIB_PREFIX}/gcc/${machine}/${version}
+clang -B ${gnuc_lib} -L ${gnuc_lib} \"$@\"' > builddir/clc"
+      system "echo '#!/bin/bash
+machine=$(gcc -dumpmachine)
+version=$(gcc -dumpversion)
+cxx_sys=#{CREW_PREFIX}/include/c++/${version}
+cxx_inc=#{CREW_PREFIX}/include/c++/${version}/${machine}
+gnuc_lib=#{CREW_LIB_PREFIX}/gcc/${machine}/${version}
+clang++ -fPIC  -rtlib=compiler-rt -stdlib=libc++ -cxx-isystem ${cxx_sys} -I ${cxx_inc} -B ${gnuc_lib} -L ${gnuc_lib} \"$@\"' > builddir/clc++"
```

Comment:
Convert wrapper scripts to use `<<~EOF ... EOF` instead of `system "echo ...`?

Reply:
Good idea.